File: ios/Sources/SignKeypair/EcdsaSignatureCodec.swift
import Foundation

/// ASN.1 DER <-> IEEE P1363 conversion for P-256 ECDSA signatures.
///
/// Deliberately free of any Tauri or Security-framework import: this type is
/// pure `Foundation`, so it can be unit-tested by a plain SwiftPM package with
/// no simulator, no Tauri runtime and no keychain. See `darwin_tests/`.
///
/// `SecKeyCreateSignature` emits **ASN.1 DER** `SEQUENCE { INTEGER r, INTEGER s }`.
/// JWS ES256 (RFC 7515 §3.4 / RFC 7518 §3.4) requires **IEEE P1363** — the two
/// integers as fixed-width 32-byte big-endian values, concatenated, unwrapped.
///
/// Getting this wrong is silent and intermittent. DER encodes integers minimally,
/// so a component whose top byte happens to be zero is one byte shorter — which
/// happens for roughly 1 signature in 256 per component. Code that copies the
/// magnitude without left-padding produces a signature that is valid ECDSA and
/// that a JWS verifier rejects, days or weeks after it shipped.
public enum EcdsaSignatureCodec {

  /// Byte width of one P-256 coordinate.
  public static let coordinateLength = 32

  /// Total length of a P-256 P1363 signature.
  public static let p1363Length = coordinateLength * 2

  /// Why a DER signature could not be decoded. Never carries key material.
  public enum DecodingError: Error, Equatable {
    /// The input ended before the structure did.
    case truncated(String)
    /// A tag byte was not the one DER requires here.
    case unexpectedTag(expected: UInt8, actual: UInt8)
    /// A declared length does not agree with the buffer.
    case lengthMismatch(String)
    /// An integer was wider than a P-256 coordinate.
    case integerTooLong(Int)
    /// A length encoding this codec deliberately does not accept.
    case unsupportedLengthForm(String)
  }

  /// Convert an ASN.1 DER ECDSA signature to IEEE P1363 `r‖s`.
  ///
  /// The result is **always** exactly ``p1363Length`` bytes, or the call throws.
  public static func derToP1363(_ der: Data) throws -> Data {
    // Index through a contiguous copy: `Data` slices keep their parent's
    // indices, and an off-by-one there is exactly the kind of out-of-bounds
    // read this codec must not have.
    let bytes = [UInt8](der)
    var offset = 0

    func readByte(_ what: String) throws -> Int {
      guard offset < bytes.count else {
        throw DecodingError.truncated("expected \(what) at offset \(offset)")
      }
      defer { offset += 1 }
      return Int(bytes[offset])
    }

    let tag = try UInt8(readByte("SEQUENCE tag"))
    guard tag == 0x30 else {
      throw DecodingError.unexpectedTag(expected: 0x30, actual: tag)
    }

    var sequenceLength = try readByte("SEQUENCE length")
    if sequenceLength & 0x80 != 0 {
      let lengthOctets = sequenceLength & 0x7f
      guard lengthOctets >= 1, lengthOctets <= 2 else {
        throw DecodingError.unsupportedLengthForm(
          "SEQUENCE length uses \(lengthOctets) octets")
      }
      sequenceLength = 0
      for _ in 0..<lengthOctets {
        sequenceLength = (sequenceLength << 8) | (try readByte("SEQUENCE length octet"))
      }
    }
    guard offset + sequenceLength == bytes.count else {
      throw DecodingError.lengthMismatch(
        "SEQUENCE declares \(sequenceLength) bytes, buffer has \(bytes.count - offset)")
    }

    func readInteger(_ what: String) throws -> ArraySlice<UInt8> {
      let tag = try UInt8(readByte("\(what) INTEGER tag"))
      guard tag == 0x02 else {
        throw DecodingError.unexpectedTag(expected: 0x02, actual: tag)
      }
      let length = try readByte("\(what) INTEGER length")
      guard length & 0x80 == 0 else {
        throw DecodingError.unsupportedLengthForm("\(what) INTEGER uses long-form length")
      }
      guard length > 0 else {
        throw DecodingError.lengthMismatch("\(what) INTEGER has zero length")
      }
      guard offset + length <= bytes.count else {
        throw DecodingError.truncated(
          "\(what) INTEGER declares \(length) bytes, only \(bytes.count - offset) remain")
      }
      defer { offset += length }
      return bytes[offset..<(offset + length)]
    }

    /// Strip DER's minimal-encoding leading zeros, then left-pad to 32 bytes.
    ///
    /// Both halves matter: a high-bit component carries a 0x00 sign byte that
    /// must NOT survive into P1363, and a short component must be padded on the
    /// LEFT — right-alignment silently multiplies the value by 256^n.
    func normalise(_ value: ArraySlice<UInt8>, _ what: String) throws -> [UInt8] {
      var start = value.startIndex
      while start < value.endIndex - 1 && value[start] == 0 {
        start += 1
      }
      let magnitude = value[start...]
      guard magnitude.count <= coordinateLength else {
        throw DecodingError.integerTooLong(magnitude.count)
      }
      var out = [UInt8](repeating: 0, count: coordinateLength)
      out.replaceSubrange((coordinateLength - magnitude.count)..<coordinateLength, with: magnitude)
      return out
    }

    let r = try normalise(try readInteger("r"), "r")
    let s = try normalise(try readInteger("s"), "s")

    // r and s must account for the entire SEQUENCE. Without this, trailing
    // bytes are silently ignored, which makes the encoding malleable: an
    // attacker can append junk and get a different byte string that this codec
    // maps to the same signature.
    guard offset == bytes.count else {
      throw DecodingError.lengthMismatch(
        "\(bytes.count - offset) trailing byte(s) after s")
    }

    let result = Data(r + s)
    // Unreachable by construction — `normalise` always returns exactly
    // `coordinateLength` bytes — and mutation testing confirms it: deleting
    // this guard fails no test. It stays as an invariant against a future edit
    // to `normalise`, because the caller ships this straight into a JWS where a
    // wrong length is a 401 with no useful diagnostics. Recorded here so the
    // next person does not mistake "untested" for "untestable".
    guard result.count == p1363Length else {
      throw DecodingError.lengthMismatch("produced \(result.count) bytes, expected \(p1363Length)")
    }
    return result
  }

  /// Convert an IEEE P1363 `r‖s` signature to ASN.1 DER.
  ///
  /// Not used on the signing path — the platform already gives us DER — but it
  /// is what lets a test round-trip against a DER-only verifier.
  public static func p1363ToDer(_ p1363: Data) throws -> Data {
    guard p1363.count == p1363Length else {
      throw DecodingError.lengthMismatch(
        "expected \(p1363Length) bytes, got \(p1363.count)")
    }
    let bytes = [UInt8](p1363)
    let body =
      derInteger(Array(bytes[0..<coordinateLength]))
      + derInteger(Array(bytes[coordinateLength..<p1363Length]))
    return Data([0x30, UInt8(body.count)] + body)
  }

  private static func derInteger(_ value: [UInt8]) -> [UInt8] {
    var start = 0
    while start < value.count - 1 && value[start] == 0 {
      start += 1
    }
    var magnitude = Array(value[start...])
    // DER INTEGERs are signed: prepend 0x00 so a high top bit is not negative.
    if magnitude[0] & 0x80 != 0 {
      magnitude.insert(0, at: 0)
    }
    return [0x02, UInt8(magnitude.count)] + magnitude
  }
}

File: android/src/main/java/app/vaam/signkeypair/EcdsaSignatureCodec.kt
package app.vaam.signkeypair

import java.math.BigInteger

/**
 * ASN.1 DER -> IEEE P1363 conversion for P-256 ECDSA signatures, plus the
 * coordinate padding the public JWK needs.
 *
 * Deliberately free of any Android import, so a plain JVM unit test can drive
 * it without a device or an emulator — `./gradlew test`, seconds rather than
 * minutes.
 *
 * `java.security.Signature` emits **ASN.1 DER** `SEQUENCE { INTEGER r, INTEGER s }`.
 * JWS ES256 (RFC 7515 §3.4 / RFC 7518 §3.4) requires **IEEE P1363** — the two
 * integers as fixed-width 32-byte big-endian values, concatenated, unwrapped.
 *
 * Getting this wrong is silent *and intermittent*. DER encodes integers
 * minimally, so a component whose top byte happens to be zero is one byte
 * shorter — roughly 1 signature in 256 per component. Code that copies the
 * magnitude without left-padding produces a signature that is valid ECDSA and
 * that a JWS verifier rejects, days or weeks after it shipped.
 */
internal object EcdsaSignatureCodec {

    /** Byte width of one P-256 coordinate. */
    const val COORDINATE_LENGTH = 32

    /** Total length of a P-256 P1363 signature. */
    const val P1363_LENGTH = COORDINATE_LENGTH * 2

    /**
     * Left-pad / trim a BigInteger to exactly 32 bytes (P-256 coordinate width).
     *
     * Same bug class as [derToP1363]: `BigInteger.toByteArray()` prepends a 0x00
     * sign byte when the high bit is set (33 bytes), and returns FEWER than 32
     * bytes for a coordinate with leading zeros. Getting either wrong corrupts
     * the public JWK sent at device registration, and the symptom is "every
     * signature this device makes is rejected" — with no hint that the *key*,
     * not the signing, was wrong.
     */
    fun coordinateBytes(value: BigInteger): ByteArray {
        val raw = value.toByteArray()
        val out = ByteArray(COORDINATE_LENGTH)
        if (raw.size > COORDINATE_LENGTH) {
            System.arraycopy(raw, raw.size - COORDINATE_LENGTH, out, 0, COORDINATE_LENGTH)
        } else {
            System.arraycopy(raw, 0, out, COORDINATE_LENGTH - raw.size, raw.size)
        }
        return out
    }

    /**
     * Convert an ASN.1 DER ECDSA signature to IEEE P1363 `r‖s`.
     *
     * The result is **always** exactly [P1363_LENGTH] bytes, or the call throws
     * [IllegalArgumentException].
     */
    fun derToP1363(der: ByteArray): ByteArray {
        var offset = 0
        fun readByte(what: String): Int {
            require(offset < der.size) { "Truncated DER signature: expected $what at offset $offset" }
            return der[offset++].toInt() and 0xff
        }

        require(readByte("SEQUENCE tag") == 0x30) { "DER signature does not start with SEQUENCE" }
        var sequenceLength = readByte("SEQUENCE length")
        if (sequenceLength and 0x80 != 0) {
            val lengthOctets = sequenceLength and 0x7f
            require(lengthOctets in 1..2) { "Unsupported DER length form ($lengthOctets octets)" }
            sequenceLength = 0
            repeat(lengthOctets) {
                sequenceLength = (sequenceLength shl 8) or readByte("SEQUENCE length octet")
            }
        }
        require(offset + sequenceLength == der.size) { "DER length mismatch" }

        fun readInteger(what: String): ByteArray {
            require(readByte("$what INTEGER tag") == 0x02) { "Expected DER INTEGER for $what" }
            val length = readByte("$what INTEGER length")
            require(length and 0x80 == 0) { "Unsupported long-form DER INTEGER for $what" }
            require(length > 0) { "Zero-length DER INTEGER for $what" }
            require(offset + length <= der.size) { "Truncated DER INTEGER for $what" }
            val value = der.copyOfRange(offset, offset + length)
            offset += length
            return value
        }

        /**
         * Strip DER's minimal-encoding leading zeros, then left-pad to 32 bytes.
         *
         * Both halves matter: a high-bit component carries a 0x00 sign byte that
         * must NOT survive into P1363, and a short component must be padded on
         * the LEFT — right-alignment silently multiplies the value by 256^n.
         */
        fun normalise(value: ByteArray): ByteArray {
            var start = 0
            while (start < value.size - 1 && value[start].toInt() == 0) start++
            val length = value.size - start
            require(length <= COORDINATE_LENGTH) { "ECDSA integer longer than $COORDINATE_LENGTH bytes" }
            val out = ByteArray(COORDINATE_LENGTH)
            System.arraycopy(value, start, out, COORDINATE_LENGTH - length, length)
            return out
        }

        val r = normalise(readInteger("r"))
        val s = normalise(readInteger("s"))

        // r and s must account for the entire SEQUENCE. Without this, trailing
        // bytes are silently ignored, which makes the encoding malleable: junk
        // can be appended to produce a different byte string that decodes to the
        // same signature. `openssl asn1parse` rejects such input too.
        require(offset == der.size) { "${der.size - offset} trailing byte(s) after s" }

        return r + s
    }
}
